stackrox/stackrox
# ROX-14526: Add exponential backoff to clarify's GetNodeVulnerabilities()

## Description

Add retries with exponential backoff when handling node scans and the Scanner is unavailable.
 
We want to retry this API call because the time for waiting for a new Node Inventory is currently high. Short availability issues on this API could cause customers to wait hours to see the latest Node Scanning components and vulnerabilities.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

1. Deployed to Openshift-4.
2. Enabled `ROX_RHCOS_NODE_SCANNING=true` in Scanner, Central, Sensor, and Collector (in that order).
3. Observe Node Scanning results in Vulnerability Management.
4. Add a network policy to block ingress on Scanner (see below).
5. Trigger a new node scan in Compliance (by kicking the container: `kubectl -n stackrox exec collector-<id> compliance -- kill 1`.
6. Observe logs.

Logs observed:

```
sensor/service/pipeline/nodeinventory: 2023/02/22 22:04:38.322413 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-88z59-worker-c-fj9wg.c.srox-temp-dev-test.internal' ID='8a9d6ea9-c28d-4b05-ba01-12a804ba90e7'
pkg/scanners/clairify: 2023/02/22 22:04:38.337873 clairify.go:438: Warn: calling GetNodeVulnerabilities() (retrying in 512.055588ms): rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.111.175:8443: connect: connection refused"
pkg/scanners/clairify: 2023/02/22 22:04:38.851134 clairify.go:438: Warn: calling GetNodeVulnerabilities() (retrying in 435.140532ms): rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.111.175:8443: connect: connection refused"
pkg/scanners/clairify: 2023/02/22 22:04:39.286958 clairify.go:438: Warn: calling GetNodeVulnerabilities() (retrying in 1.321174225s): rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.111.175:8443: connect: connection refused"
pkg/scanners/clairify: 2023/02/22 22:04:40.613495 clairify.go:438: Warn: calling GetNodeVulnerabilities() (retrying in 939.716644ms): rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.111.175:8443: connect: connection refused"
pkg/scanners/clairify: 2023/02/22 22:04:41.554490 clairify.go:438: Warn: calling GetNodeVulnerabilities() (retrying in 2.172302528s): rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 172.30.111.175:8443: connect: connection refused"
sensor/service/pipeline/nodeinventory: 2023/02/22 22:04:48.331949 pipeline.go:86: Warn: enriching node with node inventory secured-cluster-jvdm-argilo:jvdm-argilo-88z59-worker-c-fj9wg.c.srox-temp-dev-test internal: error scanning node secured-cluster-jvdm-argilo:jvdm-argilo-88z59-worker-c-fj9wg.c.srox-tem -dev-test.internal error: Error scanning 'secured-cluster-jvdm-argilo:jvdm-argilo-88z59-worker-c-fj9w .c.srox-temp-dev-test.internal' with scanner "": rpc error: code = DeadlineExceeded desc = context de dline exceeded
```

Meaning: Calls to Scanner were failing to connect, and retries were triggered as expected. Context timed out after 10 seconds, and retries gave up, as expected.

Cluster for inspection: 

```
infractl get jvdm-argilo          
ID:          jvdm-argilo
Flavor:      openshift-4

Status:      READY
Created:     2h15m ago
URL:         https://console-openshift-console.apps.jvdm-argilo.openshift.infra.rox.systems
Lifespan:    2d remaining
```

PS.: The binaries were hotplugged in the cluster, so no image build.

Network Policy:

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: deny-scanner
spec:
  podSelector:
    matchLabels:
      app: scanner
  policyTypes:
  - Egress
```
